zcash_client_sqlite/src/wallet/init.rs

```diff
@@ -1,37 +1,425 @@
+            );",
+        )?;
+        Ok(())
+    }
+
+    fn down(&self, _transaction: &Transaction) -> Result<(), WalletMigrationError> {
+        // We should never down-migrate the first migration, as that can irreversibly
+        // destroy data.
+        panic!("Cannot revert the initial migration.");
+    }
+}
+
+struct WalletMigration1;
+
+migration!(
+    WalletMigration1,
+    "a2e0ed2e-8852-475e-b0a4-f154b15b9dbe",
+    ["bc4f5e57-d600-4b6c-990f-b3538f0bfce1"],
+    "Add support for receiving transparent UTXOs."
+);
+
+impl RusqliteMigration for WalletMigration1 {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &Transaction) -> Result<(), WalletMigrationError> {
+        transaction.execute_batch(
+            "CREATE TABLE IF NOT EXISTS utxos (
+                id_utxo INTEGER PRIMARY KEY,
+                address TEXT NOT NULL,
+                prevout_txid BLOB NOT NULL,
+                prevout_idx INTEGER NOT NULL,
+                script BLOB NOT NULL,
+                value_zat INTEGER NOT NULL,
+                height INTEGER NOT NULL,
+                spent_in_tx INTEGER,
+                FOREIGN KEY (spent_in_tx) REFERENCES transactions(id_tx),
+                CONSTRAINT tx_outpoint UNIQUE (prevout_txid, prevout_idx)
+            );",
+        )?;
+        Ok(())
+    }
+
+    fn down(&self, transaction: &Transaction) -> Result<(), WalletMigrationError> {
+        transaction.execute_batch("DROP TABLE utxos;")?;
+        Ok(())
+    }
+}
+
+struct WalletMigration2<P: consensus::Parameters> {
+    params: P,
+    seed: Option<SecretVec<u8>>,
+}
+
+impl<P: consensus::Parameters> Migration for WalletMigration2<P> {
+    fn id(&self) -> Uuid {
+        ::uuid::Uuid::parse_str("be57ef3b-388e-42ea-97e2-678dafcf9754").unwrap()
+    }
+
+    fn dependencies(&self) -> HashSet<Uuid> {
+        ["a2e0ed2e-8852-475e-b0a4-f154b15b9dbe"]
+            .iter()
+            .map(|uuidstr| ::uuid::Uuid::parse_str(uuidstr).unwrap())
+            .collect()
+    }
+
+    fn description(&self) -> &'static str {
+        "Add support for unified full viewing keys"
+    }
+}
+
+impl<P: consensus::Parameters> RusqliteMigration for WalletMigration2<P> {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &Transaction) -> Result<(), WalletMigrationError> {
+        //
+        // Update the accounts table to store ufvks rather than extfvks
+        //
+
+        transaction.execute_batch(
+            "CREATE TABLE accounts_new (
+                account INTEGER PRIMARY KEY,
+                ufvk TEXT NOT NULL,
+                address TEXT,
+                transparent_address TEXT
+            );",
+        )?;
+
+        let mut stmt_fetch_accounts =
+            transaction.prepare("SELECT account, address FROM accounts")?;
+
+        let mut rows = stmt_fetch_accounts.query(NO_PARAMS)?;
+        while let Some(row) = rows.next()? {
+            if let Some(seed) = &self.seed {
```

Comment:
We shouldn't run any of the migration if we don't have the seed:
```suggestion
        let seed = self.seed.as_ref().ok_or(WalletMigrationError::SeedRequired)?;

        transaction.execute_batch(
            "CREATE TABLE accounts_new (
                account INTEGER PRIMARY KEY,
                ufvk TEXT NOT NULL,
                address TEXT,
                transparent_address TEXT
            );",
        )?;

        let mut stmt_fetch_accounts =
            transaction.prepare("SELECT account, address FROM accounts")?;

        let mut rows = stmt_fetch_accounts.query(NO_PARAMS)?;
        while let Some(row) = rows.next()? {
```

Reply:
That's not necessarily true; for example, we don't need the seed if we're initializing a brand-new database.